frame/support/src/storage/mod.rs

```diff
@@ -811,16 +812,292 @@ mod private {
 
 	impl<T: Encode> Sealed for Vec<T> {}
 	impl<Hash: Encode> Sealed for Digest<Hash> {}
+	impl<T: Value, S: Get<usize>> Sealed for BoundedVec<T, S> {}
 }
 
 impl<T: Encode> StorageAppend<T> for Vec<T> {}
 impl<T: Encode> StorageDecodeLength for Vec<T> {}
 
-/// We abuse the fact that SCALE does not put any marker into the encoding, i.e.
-/// we only encode the internal vec and we can append to this vec. We have a test that ensures
-/// that if the `Digest` format ever changes, we need to remove this here.
+/// We abuse the fact that SCALE does not put any marker into the encoding, i.e. we only encode the
+/// internal vec and we can append to this vec. We have a test that ensures that if the `Digest`
+/// format ever changes, we need to remove this here.
 impl<Hash: Encode> StorageAppend<DigestItem<Hash>> for Digest<Hash> {}
 
+use crate::traits::Get;
+use sp_std::convert::TryFrom;
+
+/// Marker trait for types `T` that can be stored in storage as `Vec<T>`.
+pub trait Value: FullCodec + Default + sp_std::fmt::Debug + Eq + PartialEq {}
+impl<T: FullCodec + Default + sp_std::fmt::Debug + Eq + PartialEq> Value for T {}
+
+/// A bounded vector.
+///
+/// It implementations for efficient append and length decoding, as with a normal `Vec<_>`, once put
+/// into a [`StorageValue`].
+///
+/// As the name suggests, the length of the queue is always bounded. All internal operations ensure
+/// this bound is respected.
+#[derive(Encode, Decode, Default, Clone, crate::RuntimeDebugNoBound)]
+pub struct BoundedVec<T: Value, S: Get<usize>>(Vec<T>, sp_std::marker::PhantomData<S>);
+
+// NOTE: we could also implement this as:
+// impl<T: Value, S1: Get<usize>, S2: Get<usize>> PartialEq<BoundedVec<T, S2>> for BoundedVec<T, S1>
+// to allow comparison of bounded vectors with different bounds.
```

Comment:
reasons not to do that?

Reply:
I don't see any use case, and it is not aligned with how the type system works by default. The default assumption is that any type `Type<T>` which is generic is transformed to a new type per generic use of `T`. Namely, `Type<A>` and `Type<B>` or not even remotely related and are totally different types from PoV of `rustc`. Therefore, I am against establishing a relation between them unless if it is needed.